man/man1/inline.1
.\" @(#)inline.1 1.1 94/10/31 SMI; from UCB 4.3 BSD
.TH INLINE 1 "23 September 1987"
.SH NAME
inline \- in-line procedure call expander
.SH SYNOPSIS
.B /usr/lib/inline
.RB [ " \-w " ]
.RB [ " \-v " ]
.RB [ " \-o"
.IR outputfile " ]"
.RB [ " \-i"
.IR inlinefile " ] .\|.\|."
.RI [ " cpu-option " ]
.RI [ " fpu-option " ]
.IR filename .\|.\|.
.SH DESCRIPTION
.IX "inline command" "" "\fLinline\fR command" 
.IX "expand assembly-language calls in-line, \fLinline\fP"
.IX "procedure calls, assembler, expand in-line, \fLinline\fP"
.LP 
.B inline
expands assembly language calls in the indicated source files
into copies of the corresponding procedure bodies obtained from an
.IR inlinefile
specified with the
.B \-i
option.
If no 
.I inlinefile
is specified, the
source files are simply concatenated and written to the standard
output.  If no source files are specified, the input is read from
the standard input.
.LP
Inline itself is little more than a sed script.  Almost all of
the benefit produced is derived from subsequent peephole
optimization.
.SH OPTIONS
.TP
.B \-w 
Display warnings for duplicate definitions on the standard error.
.TP
.B \-v
Verbose.  Display the names of routines that were actually
in-line expanded in the sourcefile on the standard error.
.TP
.BI \-o " outputfile"
write output to the indicated file; standard output by default.
.TP
.BI \-i " inlinefile"
Read in-line code templates from 
.IR inlinefile .
.TP
.I cpu-option
Specify templates for the machine architecture of a Sun-2 or Sun-3
system.  If this option is omitted, the proper template for the
host architecture is used.  Can be one of:
.RS
.RS
.TP 10
.B \-mc68010
expand
.B .mc68010
code templates
.PD 0
.TP
.B \-mc68020
expand
.B .mc68020
code templates
.PD
.RE
.RE
.TP
.I fpu-option
Specify a floating-point processor option for a Sun-2 or Sun-3
system.  Can be one of:
.RS
.RS
.TP 10
.B \-fsoft
expand
.B .fsoft
code templates (the default)
.PD 0
.TP
.B \-fswitch
expand
.B .fswitch
code templates
.TP
.B \-fsky
expand
.B .fsky
code templates
.RB ( \-mc68010
only)
.TP
.B \-f68881
expand
.B .f68881
code templates
.RB ( \-mc68020
only)
.TP
.B \-ffpa
expand
.B .ffpa 
code templates
.RB ( \-mc68020
only)
.PD
.RE
.RE
.SH USAGE
Each 
.I inlinefile
contains one or more labeled assembly language templates of the form:
.RS
.nf
.ft I
inline-directive
\&.\|.\|.
instructions
\&.\|.\|.
\&.end
.ft R
.fi
.RE
.LP
where the
.I instructions
constitute an in-line expansion of the named routine.  An
.I "inline-directive"
is a command of the form:
.IP
.BI .inline "\ \ \ identifier" , " argsize"
.LP
This declares a block of code for the routine named by
.IR identifier ,
with
.I argsize
bytes of arguments.
.RI ( argsize
is optional on Sun-4 systems).   Calls to the named routine are replaced
by the code in the in-line template.
.br
.ne 10
.LP
For Sun-2 and Sun-3 systems, the following additional forms
are recognized:
.RS
.TP 9
.B "\&.mc68010
.IB "	identifier" , " argsize"
.PD 0
.TP
.B "\&.mc68020
.IB "	identifier" , " argsize"
.TP
.B "\&.fsoft
.IB "	identifier" , " argsize"
.TP
.B "\&.fswitch
.IB "	identifier" , " argsize"
.TP
.B "\&.fsky
.IB "	identifier" , " argsize"
.TP
.B \&.f68881
.IB "	identifier" , " argsize"
.TP
.B \&.ffpa
.IB "	identifier" , " argsize"
.DT
.PD
.RE
.LP
These forms are similar to
.BR .inline ,
with the addition of a
.SM CPU
or
.SM FPU
specification.  The template is only expanded if
the specified target system matches the value of the target
.SM CPU
or
.SM FPU
type, as determined by the command-line options, or
if none were given, by the type of the host system.
.LP
Multiple templates are permitted; matching templates after the first
are ignored.  Duplicate templates may be placed in order of decreasing
performance of the corresponding hardware; thus the most efficient
usable version will be selected.
.SS Coding Conventions for all Sun Systems
In-line templates should be coded as expansions of C-compatible
procedure calls,
with the difference that the return address cannot be depended
upon to be in the expected place, since no call instruction will
have been executed.  See 
.SM FILES,
below, for examples.
.LP
In-line templates must conform to standard Sun parameter passing
and register usage conventions, as detailed below.  They must
not call routines that violate these conventions; for example,
assembly language routines such as 
.BR setjmp (3V)
may cause problems.
.LP
Registers other than the ones mentioned below must not be
used or set.
.LP
Branch instructions in an in-line template may only transfer
to numeric labels 
.RB ( 1f ,
.BR 2b ,
and so on) defined within the in-line
template.  No other control transfers are allowed.
.LP
Only opcodes and addressing modes generated by Sun
compilers are guaranteed to work.  Binary encodings of
instructions are not supported.
.LP
.SS Coding Conventions for Sun-2 and Sun-3 Systems
Arguments are passed in 32-bit aligned memory locations
starting at
.BR sp@ .
Note that there is no return address on
the stack, since no
.B jbsr
instruction will have been executed.
.LP
Results are returned in
.B d0
or
.BR d0/d1 .
.LP
The following registers may be used as temporaries: registers
.BR a0 ,
.BR a1 ,
.BR d0 ,
and
.BR d1
on the
.SM MC68010
and
.SM MC68020;
registers
.B fp0
and
.B fp1
on the
.SM MC68881;
registers
.B fpa0
through
.B fpa3
on the Sun Floating-Point Accelerator.  No other registers may be used.
.LP
The template must delete exactly
.I argsize
bytes from the
stack.  This is to enable 
.B inline
to deal with autoincrement and
autodecrement addressing modes, which in turn are used by
.B c2
to delimit the lifetimes of stack temporaries.
.LP
The stack must not underflow the level of the last argument.
.LP
Use
.BI j cc
branch mnemonics instead of
.BI b cc.
The 
.BI b cc
ops are span limited and will fail if retargeted to a label whose
span overflows the branch displacement field.
.SS Coding Conventions for Sun-4 Systems
Arguments are passed in registers
.BR %o0-%o5 ,
followed by memory locations starting at
.BR [%sp+0x5c] .
.B %sp
is guaranteed to be 64-bit aligned.  The contents of
.B %o7
are undefined, since no call instruction will have been executed.
.LP
Results are returned in
.B %o0
or
.BR %f0/%f1 .
.LP
Registers
.B %o0-%o5
and
.B %f0-%f31
may be used as temporaries.
.LP
Integral and single-precision floating-point arguments are
32-bit aligned.
.LP
Double-precision floating-point arguments are guaranteed to
be 64-bit aligned if their offsets are multiples of 8.
.LP
Each control-transfer instruction (branches and calls) must be
immediately followed by a nop.
.LP
Call instructions must include an extra (final) argument which
indicates the number of registers used to pass parameters
to the called routine.
.LP
Note that for Sun-4 systems, the instruction following an
expanded 'call' is inserted by
.B inline
.I before
the expanded code to preserve the semantics of the call's delay slot.
.SH FILES
.PD 0
.TP 20
.B /usr/lib/inline
in-line procedure call expander
.TP
.B /usr/lib/fsoft/libm.il
in-line templates for software floating point (Sun-2 and Sun-3 only)
.TP
.B /usr/lib/fswitch/libm.il
in-line templates for switched floating point (Sun-2 and Sun-3 only)
.TP
.B /usr/lib/fsky/libm.il
in-line templates for Sky FFP (Sun-2 only)
.TP
.B /usr/lib/f68881/libm.il
in-line templates for Motorola 68881 (Sun-3 only)
.TP
.B /usr/lib/ffpa/libm.il
in-line templates for Sun FPA (Sun-3 only)
.PD
.SH WARNING
.B inline
does not check for violations of the coding conventions described
above.




lang/inline/src/inline.c
#ifndef lint
static char sccsid[] = "@(#)inline.c 1.1 94/10/31 SMI";
static char RELEASE[] = "@(#)RELEASE SunOS compilers 4.1";
#endif

#include <stdio.h>
#include <ctype.h>
#include <strings.h>
#include <sys/types.h>
#include <sys/stat.h>

/*
 *  inline [-w] [-v] [-o outputfile] [-i inlinefile]
 *	[<CPU options>] [<FPU options>] [sourcefile(s)] ...
 *
 *	This program is little more than an over-embellished sed script.
 *	It inline-expands call instructions in one or more <sourcefiles>
 *	from one or more <inlinefiles>.  If no <inlinefiles> are specified,
 *	the <sourcefiles> are simply concatenated to stdout.  If no
 *	<sourcefiles> are specified, the default input is stdin.
 *
 *	If [-w] is specified, inline displays warnings for duplicate
 *	definitions.
 *
 *	If [-v] is specified, inline displays names of routines that
 *	were actually inline-expanded in the sourcefile.
 *
 *	Inline expands the subroutine call instructions of the target
 *	machine architecture.  The distinction between target machine
 *	architectures is determined at compile time by the cpp symbol
 *	TARGET.  If TARGET == MC68000, further distinctions are made
 *	at run time, based on command line options and host machine type.
 *
 *	If TARGET == MC68000, legal <CPU options> are:
 *
 *		-mc68010	(expand .mc68010 templates)
 *		-mc68020	(expand .mc68020 templates)
 *
 *	If TARGET == MC68000, legal <FPU options> are:
 *
 *		-fsoft		(expand .fsoft templates)
 *		-fswitch	(expand .fswitch templates)
 *		-fsky		(expand .fsky templates)    [-mc68010 only]
 *		-f68881		(expand .f68881 templates)  [-mc68020 only]
 *		-ffpa		(expand .ffpa templates)    [-mc68020 only]
 *
 *	In the absence of command line options, the default CPU type is
 * 	that of the host; the default FPU type is -fsoft.
 *
 *	If TARGET == SPARC, there are no CPU or FPU options (yet).
 *
 *	Each <inlinefile> contains one or more labeled assembly language
 *	fragments of the form:
 *
 *		<inline directive>
 *		...
 *		<instructions>
 *		...
 *		.end
 *
 *	where the <instructions> constitute an inline expansion of
 *	the named routine, and <inline directive> defines the attributes
 *	of the inline code.  Legal forms of <inline directives> and their
 *	meanings are as follows:
 *
 *	If TARGET == MC68000 or SPARC:
 *
 *		.inline    <identifier>,<argcount>
 *
 *	This form declares an unrestricted inline expansion template
 *	for the routine named by <identifier>, with <argcount> bytes
 *	of arguments.  Calls to the named routine are replaced by the
 *	code in the inline template.
 *
 *	If TARGET == MC68000, the following additional forms are recognized:
 *
 *		.mc68010   <identifier>,<argcount>
 *		.mc68020   <identifier>,<argcount>
 *		.fsoft	   <identifier>,<argcount>
 *		.fswitch   <identifier>,<argcount>
 *		.fsky	   <identifier>,<argcount>
 *		.f68881	   <identifier>,<argcount>
 *		.ffpa	   <identifier>,<argcount>
 *
 *	These forms are similar to .inline, with the addition of a
 *	cpu or fpu specification.  The template is only expanded if
 *	the specified target machine matches the value of the target
 *	CPU or FPU type.
 *
 *	Multiple templates are permitted; the second and subsequent
 *	matching templates are ignored.  Duplicate templates may be
 *	placed in order of decreasing performance of the corresponding
 *	hardware; thus the most efficient usable version will be selected.
 *
 *	Inline expansion templates must follow the following rules:
 *
 *	if TARGET == MC68000:
 *
 *	-- Arguments are passed in 32-bit aligned memory locations
 *	   starting at sp@.  (Note that there is no return address on
 *	   the stack, since no jbsr instruction is executed)
 *	-- Results are returned in d0 or d0/d1.
 *	-- Registers a0/a1/d0/d1/fp0/fp1/fpa0-fpa3 may be used freely.
 *	-- The routine must delete <argsize> bytes from the stack.
 *	   (this is a hack to enable inline to deal crudely with
 *	   autoincrement/autodecrement addressing modes)
 *	-- The stack should not underflow the level of the last argument.
 *
 *	if TARGET == SPARC:
 *
 *	-- Arguments are passed in registers %o0-%o5, followed by
 *	   memory locations starting at [%sp+0x5c].  %sp is guaranteed
 *	   to be 64-bit aligned.
 *	-- Results are returned in %o0 or %f0/%f1.
 *	-- Registers %o0-%o5 and %f0-%f31 may be used freely.
 *	-- Integral and single precision floating point arguments are
 *	   32-bit aligned.
 *	-- Double precision floating point arguments are guaranteed to
 *	   be 64-bit aligned if the offset is a multiple of 8.
 *
 *	   Note that on Sparc, the instruction following the 'call'
 *	   will be inserted BEFORE the inline expansion, to preserve
 *	   semantics of the call delay slot.
 *
 *	if TARGET == MC68000 or SPARC:
 *
 *	-- Registers other than those noted above must be saved and restored
 *	   by the expanded routine.
 *
 *	Warning: inline does not check for violations of the above rules.
 */

#define MAXARGS 100

char   *filename;
int     lineno;
int     nerrors;
char	*sourcenames[MAXARGS];
int	nsources = 0;
int	verbose = 0;
int	warnings = 0;

typedef enum {
    M_undef,	/* matches nothing */
    M_inline,	/* matches anything */
    M_68010,
    M_68020,
    F_soft,
    F_switch,
    F_sky,
    F_68881,
    F_fpa,
} Machine_type;

Machine_type cpu = M_undef;
Machine_type fpu = M_undef;

struct machtype {
    char *name;
    Machine_type type;
};

struct machtype inlinetype = {
    "inline",  M_inline		   /* not a command line option */
};

#if TARGET == I386
char   *callops[] = {			   /* i386 call instruction mnemonics */
	"call",
	NULL
};
#endif TARGET == I386

#if TARGET == MC68000
char   *callops[] = {			   /* m68k call instruction mnemonics */
    "jbsr",
    "jsr",
    NULL
};

struct machtype cpu_options[] = {	   /* m68k cpu types */
    {"mc68010", M_68010},
    {"mc68020", M_68020},
    NULL
};

struct machtype fpu_options[] = {	   /* m68k fpu types */
    {"fsoft",   F_soft  },
    {"fswitch", F_switch},
    {"fsky",    F_sky   },
    {"f68881",  F_68881 },
    {"ffpa",    F_fpa   },
    NULL
};

static Machine_type cputype(/*cpu, arg*/); /* parse arg string for CPU type */
static Machine_type fputype(/*fpu, arg*/); /* parse arg string for FPU type */

#endif

#if TARGET == SPARC
char *callops[] = {		/* sparc call instruction mnemonics */
    "call",
    NULL
};
#endif

struct pats {			/* inline expansion template descriptor */
    char   *name;
    char   *replace;
    short  namelen;
    short  used;
    int    argsize;
    struct pats *left, *right;
    Machine_type mtype;		/* note: can specify cpu or fpu, but not both */
};

/*
 * DUP is an error-recovery kludge for duplicate
 * routine definitions.  The second and subsequent
 * occurrences of of an inline routine are ignored.
 */
static struct pats duplicate;
#define DUP &duplicate

struct pats *ptab = NULL;	/* table of inline template definitions */
FILE *input, *output;		/* source input & output streams */

extern char *malloc();
static char *alloc();		/* malloc with out-of-memory check */
extern int errno;
extern char *sys_errlist[];

/*VARARGS*/
extern int fprintf();

/*VARARGS1*/
static void
cmderr(s, x1, x2, x3, x4)
    char *s;
{
    (void)fprintf(stderr, "inline: ");
    (void)fprintf(stderr, s, x1, x2, x3, x4);
    nerrors++;
}

/*VARARGS1*/
static void
error(s, x1, x2, x3, x4)
    char *s;
{
    (void)fprintf(stderr, "inline: %s, line %d: ", filename, lineno);
    (void)fprintf(stderr, s, x1, x2, x3, x4);
    nerrors++;
}

/*VARARGS1*/
static void
werror(s, x1, x2, x3, x4)
    char *s;
{
    if (warnings) {
	(void)fprintf(stderr, "inline: %s, line %d: ", filename, lineno);
	(void)fprintf(stderr, s, x1, x2, x3, x4);
    }
}

/*
 * add pattern to the set of expandable routines.
 * Return a pointer to the new pattern descriptor.
 * If a duplicate name is encountered, return DUP.
 */
static struct pats *
insert(mtype, name, len)
    Machine_type mtype;
    register char  *name;
    register int len;
{
    register struct pats **pp;
    register struct pats *p;
    register int n;

    pp = &ptab;
    p = *pp;
    while (p != NULL) {
	n = strncmp(name, p->name, len);
	if (n < 0) {
	    pp = &p->left;
	} else if (n > 0) {
	    pp = &p->right;
	} else if (len < p->namelen) {
	    pp = &p->left;
	} else {
	    werror("Duplicate definition: %s (ignored)\n", p->name);
	    return DUP;
	}
	p = *pp;
    }
    /* check that machine type matches cpu or fpu */
    if (mtype != M_inline && mtype != cpu && mtype != fpu ) {
	return DUP;
    }
    p = (struct pats*)alloc((unsigned)sizeof(struct pats));
    bzero((char*)p, sizeof(struct pats));
    p->name = strncpy(alloc((unsigned)len+1), name, len);
    p->namelen = len;
    p->mtype = mtype;
    *pp = p;
    return (p);
}

/*
 * look up a name in the table of inline routines
 */
static struct pats *
lookup(name, len)
    register char  *name;
    register int    len;
{
    register int    n;
    register struct pats   *p;

    p = ptab;
    while (p != NULL) {
	n = strncmp(name, p->name, len);
	if (n < 0) {
	    p = p->left;
	} else if (n > 0) {
	    p = p->right;
	} else if (len < p->namelen) {
	    p = p->left;
	} else {
	    break;
	}
    }
    return (p);
}

char *patternbuf = NULL;
char linebuf[BUFSIZ];

#define nextch(cp,c) { c = *(cp)++; }
#define skipbl(cp,c) { while (isspace(c)) nextch(cp,c); }
#define findbl(cp,c) { while (!isspace(c)) nextch(cp,c); }

/*
 * scan constant
 * constant is passed by reference
 * function returns the number of chars scanned
 */

static int
scanconst(cp,c,const)
    register char *cp;
    register char c;
    int *const;
{
    register int value, digit;
    char *temp;

    temp = cp;
    value = 0;
    if(c == '0' && *cp == 'x') {
	nextch(cp,c);
	nextch(cp,c);
	while (isxdigit(c)) {
	    c = tolower(c);
	    digit = (isdigit(c)? c - '0' : c - 'a' + 10);
	    value = value * 16 + digit;
	    nextch(cp,c);
	}
    } else if (c == '0') {
	nextch(cp,c);
	while (c >= '0' && c <= '7') {
	    value = value * 8 + (c - '0');
	    nextch(cp,c);
	}
    } else {
	while (isdigit(c)) {
	    value = value * 10 + (c - '0');
	    nextch(cp,c);
	}
    }
    *const = value;
    return(cp-temp);
}

/*
 * look up an <inline_directive> keyword.  If the keyword is
 * recognized, return the associated cpu or fpu type; otherwise
 * return M_undef. The type of ".inline" is M_inline; it matches
 * any cpu or fpu type.
 */
static Machine_type
inline_directive(s)
    char *s;
{
    char *name;

    name = inlinetype.name;
    if (strncmp(s, name, strlen(name)) == 0) {
	return inlinetype.type;
    }
#   if TARGET == MC68000
    {
	int i;
	for (i = 0; cpu_options[i].name != NULL; i++) {
	    name = cpu_options[i].name;
	    if (strncmp(s, name, strlen(name)) == 0) {
		return ( cpu_options[i].type );
	    }
	}
	for (i = 0; fpu_options[i].name != NULL; i++) {
	    name = fpu_options[i].name;
	    if (strncmp(s, name, strlen(name)) == 0) {
		return ( fpu_options[i].type );
	    }
	}
    }
#   endif
    return M_undef;
}

/*
 * read inline expansion templates from
 * a user-supplied template file.
 */
static void
readpats(fname)
    char   *fname;
{
    register    FILE * f;
    register char   c;
    register char  *cp;
    register char  *bufp;
    struct pats *pattern;
    struct stat statbuf;
    char *name;
    int len;
    unsigned filesize;
    Machine_type temptype;

    filename = fname;
    lineno = 1;
    f = fopen(filename, "r");
    if (f == NULL) {
	cmderr("cannot open inline file '%s'\n", filename);
	return;
    }
    if (stat(filename, &statbuf) < 0) {
	cmderr("cannot stat inline file '%s': %s\n",
	    filename, sys_errlist[errno]);
	return;
    }
    filesize = statbuf.st_size;
    patternbuf = alloc(filesize + (BUFSIZ/2));
    pattern = NULL;
    while (fgets(linebuf, sizeof(linebuf), f)) {
	lineno++;
	cp = linebuf;
	nextch(cp,c);
	skipbl(cp,c);
	if (c == '.') {
	    temptype = inline_directive(cp);
	    if ( temptype != M_undef ) {
		/* <inline directive> starts a new routine */
		if (pattern != NULL) {
		    error("inline routines cannot be nested\n");
		    continue;
		}
		/* skip keyword and white space */
		findbl(cp,c)
		skipbl(cp,c);
		/* scan routine name */
		if (c == '_' || c == '$' || isalpha(c)) {
		    name = cp - 1;
		    nextch(cp,c);
		    while (c == '_' || c == '$' || isalnum(c)) {
			nextch(cp,c);
		    }
		    pattern = insert(temptype, name, cp-name-1);
		    bufp = patternbuf;
		    skipbl(cp,c);
		} else {
		    error("identifier expected\n");
		}
		/*
		 * argument count indicates how many bytes of
		 * arguments are expected by the called routine.
		 * This is a kludge to enable inline to compensate
		 * crudely for the effects of autoincrement addressing.
		 * In the normal 68000 calling sequence, arguments
		 * are popped off the stack by the caller; in an inline
		 * template, the "callee" is expected to pop them.
		 */
		if(c == ',') {
		    nextch(cp,c);
		    skipbl(cp,c);
		} else {
		    error("',' expected\n");
		}
		/* scan argument count */
		if(isdigit(c)) {
		    cp += scanconst(cp,c,&pattern->argsize);
		} else {
		    error("argument count expected\n");
		}
		continue;
	    } else if (strncmp(cp, "end", 3) == 0) {
		/* .end terminates the current routine */
		if (pattern == NULL) {
		    error(".end without matching .inline\n");
		    continue;
		}
		*bufp++ = '\0';
		len = bufp - patternbuf;
		if (pattern != DUP) {
		    /*
		     * store pattern, but only the first time;
		     * i.e., ignore duplicate definitions
		     */
		    pattern->replace =
			strncpy(alloc((unsigned)len+1), patternbuf, len);
		}
		pattern = NULL;
		continue;
	    }
	}
	/* default action: stash the line */
	if (pattern != NULL) {
	    cp = linebuf;
	    while (c = *cp++) {
		*bufp++ = c;
	    }
	}
    }
    if (pattern != NULL) {
	error(".inline without matching .end\n");
    }
    free(patternbuf);
}

/*
 * Parse a source line, look for a call instruction,
 * and return an inline template descriptor if a
 * call to an inline-expandable routine is found.
 */
static struct pats *
callinst(sourceline)
    char   *sourceline;
{
    register char *cp, **op;
    register char c;
    register char *name;
    register len;

    cp = sourceline;
    nextch(cp,c);
    skipbl(cp,c);
    if (!isalpha(c))
	return NULL;
    name = cp-1;
    while(isalnum(c))
	nextch(cp,c);
    len = cp-name-1;
    for (op = callops; *op != NULL; op++) {
	if (strncmp(*op, name, len) == 0) {
	    skipbl(cp,c);
	    name = cp-1;
	    while (isalnum(c) || c == '$' || c == '_')
		nextch(cp,c);
	    return lookup(name, cp-name-1);
	}
    }
    return (NULL);
}

#if TARGET == MC68000

/*
 * return CPU option selected by a command-line argument
 * string.  Issue a warning if arg conflicts with an earlier
 * option (i.e., cpu != M_undef)
 */
static Machine_type
cputype(cpu, arg)
    Machine_type cpu;
    char *arg;
{
    int i;

    for (i = 0; cpu_options[i].name != NULL; i++) {
	if (strcmp(cpu_options[i].name, arg+1) == 0) {
	    /*
	     * arg matches option from table.  Return
	     * the latter; string address is significant
	     */
	    if (cpu != M_undef && cpu != cpu_options[i].type) {
		cmderr("conflicting cpu option \"-%s\" ignored\n",
		    cpu_options[i].name);
		return cpu;
	    }
	    return cpu_options[i].type;
	}
    }
    return M_undef;
}

/*
 * return the FPU option selected by a command-line argument
 * string.  Similar to cputype().
 */
static Machine_type
fputype(fpu, arg)
    Machine_type fpu;
    char *arg;
{
    int i;

    for (i = 0; fpu_options[i].name != NULL; i++) {
	if (strcmp(fpu_options[i].name, arg+1) == 0) {
	    /*
	     * arg matches option from table.  Return
	     * the latter; string address is significant
	     */
	    if (fpu != M_undef && fpu != fpu_options[i].type) {
		cmderr("conflicting fpu option \"-%s\" ignored\n",
		    fpu_options[i].name);
		return fpu;
	    }
	    return fpu_options[i].type;
	}
    }
    return M_undef;
}

/*
 * default CPU type = what we're running on
 */
static Machine_type
default_cputype()
{
#   ifdef mc68020
	return M_68020;
#   endif
#   ifdef mc68010
	return M_68010;
#   endif
}

/*
 * default FPU type = fsoft
 */
static Machine_type
default_fputype()
{
    return F_soft;
}

/*
 * check that FPU and CPU types form
 * a sensible combination
 */
static void
check_fputype()
{
    if (cpu == M_undef)
	cpu = default_cputype();
    if (fpu == M_undef)
	fpu = default_fputype();
    if (cpu == M_68010 && fpu == F_fpa
      || cpu == M_68010 && fpu == F_68881
      || cpu == M_68020 && fpu == F_sky) {
	cmderr("%s and %s are not a supported combination\n", cpu, fpu);
    }
}

/*
 * emit code to adjust the stack pointer following
 * the inline expansion, to  compensate for the fact
 * that the template (presumably) has deleted the arguments
 * from the stack (which usually is done by the caller).
 *
 * 1e+308 on the vomit meter.
 */
static void
adjust(pattern, output)
    struct pats *pattern;
    FILE *output;
{
    int n;

    n = pattern->argsize;
    if (n > 0) {
	(void)fprintf(output, "	%s	#%d,sp\n",
	    n <= 8 ? "subqw" : "subw", n);
    } else if (n < 0) {
	error(".inline routine %s has illegal argument size (%d)\n",
	    pattern->name, n);
    }
}
#endif MC68000

#if TARGET == I386
/* see comment for MC68000 version */
static void
adjust(pattern, output)
    struct pats *pattern;
    FILE *output;
{
    int n;

    n = pattern->argsize;
    if (n > 0) {
	(void)fprintf(output, "	subl	$%d,\%%esp\n",n);
    } else if (n < 0) {
	error(".inline routine %s has illegal argument size (%d)\n",
	    pattern->name, n);
    }
}
#endif TARGET == I386

static void
summary(p)
    register struct pats *p;
{
    if (p != NULL) {
	summary(p->left);
	summary(p->right);
	if (p->used) {
	    (void)fprintf(stderr,"\t%s\n", p->name);
	}
    }
}

static void
expand(input,output)
    register FILE *input;
    register FILE *output;
{
    register struct pats *pattern;

    while (fgets(linebuf, sizeof(linebuf), input)) {
	pattern = callinst(linebuf);
	if (pattern == NULL || pattern->replace == NULL) {
	    fputs(linebuf, output);
	} else {
	    pattern->used = 1;
#	    if TARGET == SPARC
		/* write out the "delay slot" instruction first */
		if (fgets(linebuf, sizeof(linebuf), input) == NULL) {
		    error("no delay slot in call");
		}
		fputs(linebuf,output);
#	    endif
	    fputs(pattern->replace, output);
#	    if (TARGET == MC68000) || (TARGET == I386)
		adjust(pattern, output);
#	    endif
	}
    }
    if (verbose) {
	(void)fprintf(stderr,"Expanded:\n");
	summary(ptab);
    }
}

static void
init(argc, argv)
    int argc;
    char *argv[];
{
    char *arg;
    int i;

    output = stdout;
    for (i = 1; i < argc; i++) {
	arg = argv[i];
	if (arg[0] == '-') {
	    switch(arg[1]) {
	    case 'i':
		/* argv[i+1] is the name of an inline file */
		if (i+1 < argc) {
		    readpats(argv[i+1]);
		    i++;
		} else {
		    cmderr("filename expected after -i\n");
		}
		break;
	    case 'o':
		/* argv[i+1] is the name of an output file */
		if (i+1 < argc) {
		    i++;
		    output = fopen(argv[i], "w");
		    if (output == NULL) {
			cmderr("cannot open output file '%s'\n", argv[i]);
			output = stdout;
		    }
		} else {
		    cmderr("filename expected after -o\n");
		}
		break;
	    case 'v':
		verbose++;
		break;
	    case 'w':
		warnings++;
		break;
#	    if TARGET == MC68000
		case 'm':
		    cpu = cputype(cpu, arg);
		    break;
		case 'f':
		    fpu = fputype(fpu, arg);
		    break;
#	    endif MC68000
	    default:
		cmderr("unknown option (%s)\n", arg);
		break;
	    }
	} else {
	    /* arg is the name of a source file */
	    if (nsources < MAXARGS) {
		sourcenames[nsources++] = arg;
	    } else {
		cmderr("too many arguments\n");
		break;
	    }
	}
    }
}

main(argc, argv)
    int argc;
    char *argv[];
{
    int i;

    init(argc, argv);
#   if TARGET == MC68000
	check_fputype();
#   endif
    if (nsources) {
	for(i = 0; i < nsources; i++) {
	    /* process source files in sequence */
	    input = fopen(sourcenames[i], "r");
	    if (input == NULL) {
		cmderr("cannot open source file '%s'\n", sourcenames[i]);
		continue;
	    }
	    expand(input, output);
	}
    } else {
	/* source defaults to standard input */
	input = stdin;
	expand(input, output);
    }
    exit(nerrors);
}

static char *
alloc(n)
    unsigned n;
{
    char *p;

    p = malloc(n);
    if (p == NULL) {
	perror("inline");
	exit(1);
    }
    return p;
}
